evm/src/memory/memory_stark.rs

```diff
@@ -458,28 +463,33 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for MemoryStark<F
         let range_check_diff = builder.sub_extension(range_check, computed_range_check);
         yield_constr.constraint_transition(builder, range_check_diff);
 
-        // Enumerate purportedly-ordered log.
+        // Validate new_first_read. It's 1 iff the address changed and the first operation at the new address is a read.
+        let new_first_read = local_values[NEW_FIRST_READ];
+        let computed_new_first_read = builder.mul_extension(not_address_unchanged, next_is_read);
+        let new_first_read_constraint =
+            builder.sub_extension(new_first_read, computed_new_first_read);
+        yield_constr.constraint_transition(builder, new_first_read_constraint);
+
         // We assume that memory is initialized with 0. This means that if the first operation of a new address
         // is a read, then its value must be 0.
         for i in 0..8 {
+            // Enumerate purportedly-ordered log.
             let value_diff = builder.sub_extension(next_values_limbs[i], value_limbs[i]);
             let zero_if_read = builder.mul_extension(address_unchanged, value_diff);
             let read_constraint = builder.mul_extension(next_is_read, zero_if_read);
             yield_constr.constraint_transition(builder, read_constraint);
-            let first_read_value =
-                builder.mul_extension(next_values_limbs[i], not_address_unchanged);
-            let first_read_constraint = builder.mul_extension(first_read_value, next_is_read);
-            yield_constr.constraint_transition(builder, first_read_constraint);
+            // For any segment in non-zero contexts, we assume that memory is initialized with 0. This means that if
+            // the first operation of a new address is a read, then its value must be 0.
+            let first_read_value = builder.mul_extension(next_values_limbs[i], new_first_read);
+            let first_read_user_context_constraint =
+                builder.mul_extension(next_addr_context, first_read_value);
+            yield_constr.constraint_transition(builder, first_read_user_context_constraint);
+            // For any segment outside the code segment, memory is initialized with 0. Note that the previous constraint
+            // ensure that only the code segment at context 0 is not zero-initialized.
+            let kernel_context_initializing_constraint =
+                builder.mul_extension(next_addr_segment, first_read_value);
+            yield_constr.constraint_transition(builder, kernel_context_initializing_constraint);
```

Comment:
This is fine for now, but ideally I think we should allow non-zero initialized memory for all code segments. That way we wouldn't have to load contract code from prover inputs. Obviously we would need to take care of some security issues when doing that.

Reply:
Some must be zero-initialized for soundness (typically the main memory segment of any non-zero context), so I'd very much rather have all segments zero-initialized by default, with specific exceptions (like the code segment) made, as long as the constraints don't become too complicated and it doesn't add a prohibitive amount of extra columns.

Reply:
Right, I was talking about the code segments, i.e., `segment=0`. If I'm correctly understanding your constraints, this allows non-zero initialized memory only for the kernel code segment `context=0, segment=0`.

Reply:
Oh yeah, sorry. Indeed, we should unconstrain the code segment for all contexts in the future.